src/mol-math/geometry/symmetry-operator.ts

```diff
@@ -167,127 +168,111 @@ namespace SymmetryOperator {
     export interface ArrayMapping<T extends number> {
         readonly coordinates: Coordinates,
         readonly operator: SymmetryOperator,
-        readonly invariantPosition: CoordinateMapper<T>,
-        readonly position: CoordinateMapper<T>,
-        x(index: T): number,
-        y(index: T): number,
-        z(index: T): number,
-        r(index: T): number
+        invariantPosition(this: ArrayMapping<T>, i: T, s: Vec3): Vec3,
+        position(this: ArrayMapping<T>, i: T, s: Vec3): Vec3,
+        x(this: ArrayMapping<T>, index: T): number,
+        y(this: ArrayMapping<T>, index: T): number,
+        z(this: ArrayMapping<T>, index: T): number,
+        readonly r: (this: ArrayMapping<T>, index: T) => number
     }
 
-    export interface Coordinates { x: ArrayLike<number>, y: ArrayLike<number>, z: ArrayLike<number> }
+    class _ArrayMapping<T extends number> implements ArrayMapping<T> {
+        constructor(readonly operator: SymmetryOperator, readonly coordinates: Coordinates, readonly r: ((index: T) => number) = _zeroRadius) { }
 
-    function _createMapping<T extends number>(operator: SymmetryOperator, coords: Coordinates, radius: ((index: T) => number)): ArrayMapping<T> {
-        const invariantPosition = createCoordinateMapper(SymmetryOperator.Default, coords);
-        const position = operator.isIdentity ? invariantPosition : createCoordinateMapper(operator, coords);
-        const { x, y, z } = createProjections(operator, coords);
-        return { operator, coordinates: coords, invariantPosition, position, x, y, z, r: radius };
-    }
+        invariantPosition(i: T, s: Vec3): Vec3 {
+            s[0] = this.coordinates.x[i];
+            s[1] = this.coordinates.y[i];
+            s[2] = this.coordinates.z[i];
+            return s;
+        }
 
-    export function createMapping<T extends number>(operator: SymmetryOperator, coords: Coordinates, radius: ((index: T) => number) = _zeroRadius) {
-        return _createMapping(operator, coords, radius);
-    }
+        position(i: T, s: Vec3): Vec3 {
+            s[0] = this.coordinates.x[i];
+            s[1] = this.coordinates.y[i];
+            s[2] = this.coordinates.z[i];
+            Vec3.transformMat4(s, s, this.operator.matrix);
+            return s;
+        }
 
-    export function createCoordinateMapper<T extends number>(t: SymmetryOperator, coords: Coordinates): CoordinateMapper<T> {
-        if (t.isIdentity) return identityPosition(coords);
-        return generalPosition(t, coords);
-    }
-}
+        x(i: T): number {
+            const m = this.operator.matrix;
+            const { x: xs, y: ys, z: zs } = this.coordinates;
+            const xx = m[0], yy = m[4], zz = m[8], tx = m[12];
 
-export { SymmetryOperator };
+            const x = xs[i], y = ys[i], z = zs[i], w = (m[3] * x + m[7] * y + m[11] * z + m[15]) || 1.0;
+            return (xx * x + yy * y + zz * z + tx) / w;
+        }
```

Comment:
Have you tested the performance difference as well? The object deconstruction here and `this.` accesses were not here before.

This one I am not sure about, but if we go to ES6 and the deconstruction (`{ x: xs, y: ys, z: zs }`) isn't transpiled to assignment, that could be an issue too (actually, not sure that is happening now either and their isn't just some helper method used to unpack the object).

Reply:
did some perf tests, see other comment